Gravity announcement (#2643)
mainmatter/mainmatter.com: src/posts/2025-03-19-gravity.md

---
title: "Introducing Gravity: An Easier Way to Keep Build Artifact Sizes Under Control"
authorHandle: oscard0m_
tags: [gravity, ci, bundlesize, performance]
bio: "Oscar Dominguez, Software Engineer"
description: "Gravity monitors build artifact sizes to prevent accidental increases – right in your CI pipeline."
og:
  image: /assets/images/posts/2025-03-19-gravity/og-image.jpg
image: "/assets/images/posts/2025-03-19-gravity/man-falling-darker-gray-logo.webp"
imageAlt: "Gravity logo and backgound."
---

Build artifact sizes might not be the first thing you think about during code reviews, but keeping them in check can make a big difference.

> It's easy to add a JS dependency for a tiny feature that adds megabytes to the JS bundle without noticing or adding a huge _16MB_ _8000x8000_ pixel JPEG to a website or even adding a dependency to a Rust crate that results in a huge increase of the released binary. Any minor update can easily lead to an unexpected size increase, and if it goes unnoticed, it could cause issues down the line.

Reviewer: With shorter sentences, specific examples, how about something like that:
```suggestion
It’s easy to bloat your build without realizing it. Maybe you just needed a simple date formatter and pulled in `moment.js`, unknowingly adding hundreds of kilobytes of JavaScript. Or you accidentally shipped a 16MB 8000×8000 pixel JPEG. Or maybe you added `serde_json` to a Rust project just to parse a tiny config file, only to see your binary size jump by megabytes. These small choices add up, and if they go unnoticed, they can hurt performance, user experience, and maintainability.
```

---
title: "Gravity: Keep Your Bundle Sizes under Control"
authorHandle: oscard0m_
tags: [ci, bundlesize, performance]
bio: "Oscar Dominguez, Software Engineer"
description: "TODO"
og:
  image: /assets/images/posts/2025-03-19-gravity/og-image.png
image: "/assets/images/posts/2025-03-19-gravity/man-falling-darker-gray-logo.webp"
imageAlt: "Gravity monitors build artifact sizes to prevent accidental increases – right in your CI pipeline."
---

> Tired of unexpected bloated builds slowing down your app? Meet [Gravity](https://gravity.ci) by [Mainmatter](https://mainmatter.com), your new CI/CD tool for keeping bundle sizes in check, effortlessly.

Reviewer 2: We should open the post differently, not as a sales pitch but by describing the problem with e.g. some typical scenarios:

* it's easy to add a JS dependency for a tiny feature that adds megabytes to the JS bundle without noticing – most common example I know of is https://github.com/dropbox/zxcvbn
* adding a huge 16MB 8000x8000 pixel JPEG to a website
* add a dependency to a Rust crate that results in a huge increase of the released binary

It's hard to notice these things. The workflow we've used for visual regression testing is great for this though – surfacing any changes and letting the dev decide whether they are ok or not. That's the problem that Gravity solves

Reviewer 2: Also be careful to use the words carefully – bundle means JS bundle typically but Gravity isn't only for JS bundles. I'd use artifacts but there might be a better word.

Author: > We should open the post differently, not as a sales pitch but by describing the problem with e.g. some typical scenarios:
> 
> * it's easy to add a JS dependency for a tiny feature that adds megabytes to the JS bundle without noticing – most common example I know of is https://github.com/dropbox/zxcvbn
> * adding a huge 16MB 8000x8000 pixel JPEG to a website
> * add a dependency to a Rust crate that results in a huge increase of the released binary
> 
> It's hard to notice these things. The workflow we've used for visual regression testing is great for this though – surfacing any changes and letting the dev decide whether they are ok or not. That's the problem that Gravity solves

I will try again, let's see if I get the angle we want to give. Thanks for the feedback.

## What is Gravity?

_Gravity_ automatically tracks build size changes in your pull request. If your bundle gets bigger, _Gravity_ flags it for approval. Otherwise, you can merge with confidence.

This ensures that accidental increases in asset or binary sizes don’t hit production, helping you maintain the performance of your applications together with a good user experience.

### Why We Built It?

Performance regressions sneak in when you least expect them. With teams shipping updates multiple times a day, keeping an eye on bundle size growth can feel impossible. But ignoring it leads to slow load times and a frustrating user experience.

That’s why we built _Gravity_, a smooth way for product teams to stay on top of performance without tedious manual reviews.

<!-- TODO: Find an external resource on what happens when you have very large bundle sizes in your websites -->

Here are some cases where you might want to use _Gravity_:

> - Catch unminified HTML/CSS/JS files

Reviewer 2: We don't specifically catch unminified files

Author: We don't detect that directly, but one of the reasons for a bigger file size could be a non-minified HTML/CSS (let's say someone messes up with the `vite` config).

Reviewer 2: saying this like this is very misleading though

- Catch images not properly compressed
- Prevent a new version of your library from being published that is too large
- Avoid unexpected bloated bundles when installing an update of a dependency

> - Catch dead code you don't need to publish

Reviewer 2: We can't detect dead code

Author: We don't detect that directly, but one of the reasons for a bigger file size could be not tree-shaken correctly (let's say a named import is replaced by a default import and not treeshaken correctly for some reason).

Reviewer 2: saying this like this is very misleading though

> - Catch dead code you don't need to publish

Reviewer 2: I'm not sure this list is helpful tbh when we explain what kind of scenarios we catch in the beginning…

### The _bundle size_ increased 📈

In case the bundle size **increased**, the status of this check will be **failure (Needs approval) ❌**.

![Gravity Status Check needs approval](/assets/images/posts/2025-03-19-gravity/needs-approval.png)

In this case, Gravity expects the team to decide if the reported increment is expected or not. If it's not expected, the author of the pull request has to address the issue. If the increment of bytes is expected, Gravity expects a reviewer to access the Gravity Dashboard and approve the bundle increase.

By accessing the Gravity Dashboard, the reviewer can see, in a quick eye-scan, which files have increased in size, and how much.

![Approval page](/assets/images/posts/2025-03-19-gravity/approval-page.mp4#video)

Once this is done, the Gravity PR Check will become **approved** and can be merged.

![Gravity check approved](/assets/images/posts/2025-03-19-gravity/approved.png)

### Limitations

- For now we only support [GitHub Actions](https://github.com/features/actions), but we plan to add support for other CI/CD tools in the future.

> - Bundle size limit of 5GB.

Reviewer 2: Is this indeed correct?

- We don't support private repositories yet.

> - We don't support self-hosted server for now.

Reviewer: This should be rephrased in a positive way, which isn't going to be easy :D Something like:

```md
### Future plans

- Gravity already works with GitHub Actions, and we plan to support other CI/CD tools soon
- Gravity supports bundle sizes of up to 5GB (do we actually need to mention it on this blog post? Is this relevant to many users? It could go to a Q&A page at some point)
- We support open-source repos and plan on adding support for private repos next
- At this time, self-hosted servers are not supported (same as the bundle size limit: is it worth mentioning on this blog post?)

> - We don't support self-hosted server for now.

Reviewer 2: This sounds like we might support that later which I'm not sure about

Reviewer 2: no, we only needed that for our own testing

## Free for Open Source:

> We are strong believers [companies should invest in Open Source](https://mainmatter.com/blog/2019/11/11/why-companies-invest-in-oss/).
> 
> _Gravity_ is not an exception: It is **free for open-source** repositories!

Reviewer: ```suggestion
At Mainmatter, we strongly believe [companies should invest in Open Source](https://mainmatter.com/blog/2019/11/11/why-companies-invest-in-oss/). We maintain x and y, we run the [Ember Initiative]… _Gravity_ is not an exception: It is **free for open-source** repositories!
```

---
title: "Introducing Gravity: An Easier Way to Keep Build Artifact Sizes Under Control"
authorHandle: oscard0m_
tags: [gravity, ci, bundlesize, performance]
bio: "Oscar Dominguez, Software Engineer"
description: "Gravity monitors build artifact sizes to prevent accidental increases – right in your CI pipeline."
og:
  image: /assets/images/posts/2025-03-19-gravity/og-image.png

> image: "/assets/images/posts/2025-03-19-gravity/man-falling-darker-gray-logo.webp"

Reviewer 2: the image doesn't seem to work – looks very poor quality and most of it is cut off anyway

Author: Yes. It's a placeholder. We need to decide what image should go here. Already mentioned in PR's todo list.

imageAlt: "Gravity logo and backgound."
---

Build artifact sizes might not be the first thing you think about during code reviews, but keeping them in check can make a big difference. A minor update can easily lead to an unexpected size increase, and if it goes unnoticed, it could cause issues down the line.

> At [Mainmatter](https://mainmatter.com), we've seen firsthand how even small changes can lead to surprising performance regressions. That's why we built [Gravity](https://gravity.ci), a tool that helps developers manage build artifact sizes directly within their CI/CD pipeline. Today, we're excited to share it with the world.

Reviewer: Reposting Marco's comment, I think actually describing specific but common dev problems could be useful in the opener:

> - it's easy to add a JS dependency for a tiny feature that adds megabytes to the JS bundle without noticing – most common example I know of is https://github.com/dropbox/zxcvbn
> - adding a huge 16MB 8000x8000 pixel JPEG to a website
> - add a dependency to a Rust crate that results in a huge increase of the released binary

For instance, "adding a huge 16MB 8000x8000 pixel JPEG to a website" is more interesting to read than what I suggested, "an uncompressed image"

> At [Mainmatter](https://mainmatter.com), we've seen firsthand how even small changes can lead to surprising performance regressions. That's why we built [Gravity](https://gravity.ci), a tool that helps developers manage build artifact sizes directly within their CI/CD pipeline. Today, we're excited to share it with the world.

Reviewer 2: We definitely need these examples. The point to make is that all of those examples are changes that aren't **intentional** – nobody means to add a 16MB image to a blog post or a huge JS dependency. Yet, it's very hard to be aware of this since you only see e.g. a dependency being added but can't see what impact in MB that has on the bundle – that's the problem that Gravity solves, surfacing those consequences and forcing the developer to review them (which is exactly what the visual regression testing tools do for visual changes)

> ## What is Gravity?

Reviewer 2: In this part we should also talk about how we got inspiration from visual regression testing tools like Percy and Chromatic that established this workflow of 1. making the developer aware of the (in that case, visual) implications of their changes, and 2. forcing them to review those.

> Gravity is a tool that helps developers manage build artifact sizes directly within their CI/CD pipeline. It monitors all types of artifacts, including binaries, JavaScript bundles, images, CSS files, and more, by comparing changes in pull requests with the [_default branch_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-branches#about-the-default-branch).

Reviewer: ```suggestion
Gravity is a tool that helps developers manage build artifact sizes directly within their CI/CD pipeline. It monitors all types of artifacts, including binaries, JavaScript bundles, images, CSS files, and more, by comparing changes in pull requests.
```

Just trying to keep things shorter, not sure the default branch part is essential since the majority of PRs are against the default branch. Up to you!

When artifact sizes increase, Gravity flags them, providing reviewers with a clear overview to assess whether the change is expected or needs further attention. This process streamlines code reviews, reduces the risk of human error, and helps maintain the quality and performance of your projects—whether you're building applications, libraries, or tools.

## Why Use Gravity?

☑️ **Improve Review Quality**

> See exactly which files have changed in size. Gravity gives you a clear overview right where you need it—in your pull request.

Reviewer: Here as well I'd use specific engineering examples, e.g. "Someone added an image but forgot to compress it? A new dependency but the library is bloated? Some CSS isn't properly minimized? Learn about it before merging a PR"

> ☑️ **Improve Review Quality**
> 
> See exactly which files have changed in size. Gravity gives you a clear overview right where you need it—in your pull request.
> 
> ☑️ **Maintain Performance**
> 
> Identify accidental size increases early on, and catch potential performance issues before they reach production.
> 
> ☑️ **Save Time**

Reviewer: ```suggestion
☑️ **Improve review quality**

See exactly which files have changed in size. Gravity gives you a clear overview right where you need it—in your pull request.

☑️ **Maintain performance**

Identify accidental size increases early on, and catch potential performance issues before they reach production.

☑️ **Save time**
```

Avoid wasting time backtracking through commits. Gravity highlights size changes automatically, making code reviews faster and easier.

## How Does Gravity Work?

> Gravity integrates directly into your CI/CD pipeline, combining a GitHub app, a CLI tool in your GitHub Action, and a robust server-side analysis process. Once set up, Gravity automatically triggers an analysis every time a pull request is opened or modified. Here's a closer look at how it works:

Reviewer: ```suggestion
Gravity integrates directly into your CI/CD pipeline, combining a GitHub app, a CLI tool in your GitHub Actions, and a robust server-side analysis process. Once set up, Gravity automatically triggers an analysis every time a pull request is opened or modified. Here's a closer look at how it works:
```

> Gravity integrates directly into your CI/CD pipeline by combining a GitHub app, a CLI tool in your GitHub Actions, and a robust server-side analysis process. Once set up, Gravity automatically triggers an analysis every time a pull request is opened or modified. Here's a closer look at how it works:

Reviewer 2: I think a diagram might be nice here – I believe we have Mermaid set up already.

### 1. Artifact Upload

> When there are changes in a pull request, Gravity's CLI tool collects the build artifacts and uploads them to [S3](https://aws.amazon.com/s3/), and calls our server to schedule an analysis, adding you to a queue ([Redis](https://redis.io/)).

Reviewer: - I'd mention that the script needs to build artifacts before uploading them (then up to you to mention specific examples, like bundling dependencies, minimizing/compressing assets, or compiling binaries)
- Is it worth mentioning that the S3 bucket is secure? Is it even secure? I guess for open-source repos it's not really needed, though we don't want to expose the build artifacts otherwise people could just serve them from there :D
- I'd rephrase "our server" into "the Gravity server". Generally, let's be careful about "we", "us", "our". Let's use Mainmatter or Gravity where appropriate

```suggestion
When there are changes in a pull request, Gravity's CLI tool running in a GitHub Action builds your app then uploads the build artifacts to a secure [S3 bucket](https://aws.amazon.com/s3/). The CLI tool then calls the Gravity server to schedule an analysis, using a ([Redis](https://redis.io/)) queue.
```

Reviewer 2: > When there are changes in a pull request

This isn 't correct. Gravity uploads the artifacts for every CI run. If the upload is for a pull request (regardless of whether there are changes), we compare what's uploaded to the PR target. If the upload isn't for a pull request, we simply store the upload for the respective sha.

Author: >- I'd mention that the script needs to build artifacts before uploading them (then up to you to mention specific examples, like bundling dependencies, minimizing/compressing assets, or compiling binaries)

👍🏽 

> -Is it worth mentioning that the S3 bucket is secure? 

This is a FAQ question and not something key for this blog post in my opinion. According to AWS S3 docs yes, it's secure.

> Is it even secure? I guess for open-source repos it's not really needed, though we don't want to expose the build artifacts otherwise people could just serve them from there :D

As far as I know, you can't access our S3 resources unless you have the `ACCESS_KEY_ID` and `SECRET_ACCESS_KEY`.

> I'd rephrase "our server" into "the Gravity server". Generally, let's be careful about "we", "us", "our". Let's use Mainmatter or Gravity where appropriate

👍🏽

Author: > > When there are changes in a pull request

> This isn 't correct. Gravity uploads the artifacts for every CI run. If the upload is for a pull request (regardless of whether there are changes), we compare what's uploaded to the PR target. If the upload isn't for a pull request, we simply store the upload for the respective sha.

👍🏽

### 2. Server-Side Analysis

> When it’s your turn in the queue, the analysis starts, comparing the artifact sizes in the pull request with those in the [_default branch_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-branches#about-the-default-branch). Gravity does **not** inspect the content of the files, focusing purely on size changes to maintain data privacy and security.

Reviewer: - Here as well, "your turn in the queue" vs. "the PR's turn", although I wouldn't mention it as it implies wait time.
- I think I commented this before but I don't see much value in adding a link about default branches. The PR is against a branch, the developers know that, and Gravity should probably compare the PR vs. its base branch (not always the default branch). Too many links can be distracting
- Since the content of the files is not inspected, why do we even upload them? Couldn't the CLI build that report about the build (files, paths and sizes) and upload that to S3? It's already reading everything since it's uploading them, so it would be faster, lighter and more secure

Reviewer: Also this section, I'd mention some tech we use on the backend, some interesting constraints (Rust, in-memory instead of unzipping etc.)

Reviewer 2: > with those in the [_default branch_]

This isn't correct. We compare against the PR target which might not be the default branch

Author: > - Here as well, "your turn in the queue" vs. "the PR's turn", although I wouldn't mention it as it implies wait time.

👍🏽 

> - I think I commented this before but I don't see much value in adding a link about default branches. The PR is against a branch, the developers know that, and Gravity should probably compare the PR vs. its base branch (not always the default branch). Too many links can be distracting

removed 👍🏽 

> - Since the content of the files is not inspected, why do we even upload them? Couldn't the CLI build that report about the build (files, paths and sizes) and upload that to S3? It's already reading everything since it's uploading them, so it would be faster, lighter and more secure

These are FAQs questions to me. We can create a gravity.ci/faq (I see a lot of value in it), but replying to these questions in the blog post can be overwhelming to the reader.

### 3. Real-Time Feedback in GitHub

Once the analysis is finished, Gravity updates the [_GitHub status check_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/about-status-checks) directly within the pull request.

> **3.1.** If the artifact size has not increased, the status check is automatically approved, and the pull request can be merged without additional steps.

Reviewer: I think there's a slight nuance between the check status and whether the PR can be merged. Up to you to get into the details in this post, but in my opinion:
- green check: the developer can merge the PR with peace of mind, knowing that artifact sizes didn't grow
- red check: the developer can still choose to merge the PR (either manually overriding GH, or setting up GH so that the check isn't required), after checking Gravity or not at all. The developer can also set up GH to require a green check and get the peace of mind that _someone_ has to manually confirm that file size increases are acceptable

![Screenshot of Gravity Check in GitHub: Automatically approved](/assets/images/posts/2025-03-19-gravity/automatically-approved.png)

**3.2.** If Gravity detects a size increase, the status check will fail. Reviewers are prompted to access the Gravity Dashboard, where they can see a detailed breakdown of which files increased in size and by how much.

![Screenshot of Gravity Check in GitHub: It needs approval](/assets/images/posts/2025-03-19-gravity/needs-approval.png)

![Approval page](/assets/images/posts/2025-03-19-gravity/approval-page.mp4#video)

If the reviewer deems the size increase acceptable, they can approve the change directly in the Gravity Dashboard. This approval updates the status check in GitHub, allowing the pull request to proceed to merge.

![Screenshot of Gravity Check in GitHub: Approved](/assets/images/posts/2025-03-19-gravity/approved.png)

<p style="color:red">[INSERT VIDEO HERE]<p>

> ## See Gravity in Action

Reviewer 2: I think this is a repetition of what we have at the top already. I'd remove it here.

Reviewer 2: Instead of this, I'd go into a bit more technical detail, saying e.g. what we're using to build Gravity (TS, TypeORM, Rust via NAPI), how we host this. etc.

> Here are 3 Development Challenges Solved by Gravity:

Reviewer: This sentence doesn't need capitalization, feels very ChatGPT

Author: ```suggestion
Here are 3 development challenges solved by Gravity:
```

## How does Gravity work?

Inspired by visual regression testing tools like [Percy](https://percy.io/) and [Chromatic](https://www.chromatic.com/), Gravity integrates seamlessly into the developer workflow:

1. Gravity analyzes production build artifacts for every CI run.
2. For each new pull request, Gravity compares the artifacts generated by the PR against those from the target branch. If it detects new or growing artifacts, it adds a failing check to the PR, preventing it from being merged.
3. The developer reviews the changes in Gravity. If the changes are intentional and appropriate, they approve them. Otherwise, they address the issues.
4. Once all changes are approved, the Gravity check turns green, and the PR is ready to merge.

Here's a brief video of Gravity in action:

TODO: add video

## How Does Gravity Work?

Gravity consists of a set of core components:

- the CLI tool [`@gravityci/cli`](https://www.npmjs.com/package/@gravityci/cli), which runs in your CI environment
- The Gravity server, which processes uploaded artifacts, calculates differences, and hosts the UI for developers to review and approve changes.

> - the GitHub[^1], which enables Gravity to interact with the repository, such as adding checks to pull requests.

Author: ```suggestion
- the GitHub app[^1], which enables Gravity to interact with the repository, such as adding checks to pull requests.
```
